recommenders/evaluation/spark_evaluation.py

```diff
@@ -588,6 +608,15 @@ def _get_pairwise_items(self, df):
         )
 
     def _get_cosine_similarity(self, n_partitions=200):
+        if self.item_feature_df is None:
```

Comment:
It would be better to select the method for calculating similarity using an argument instead of whether `item_feature_df` is present or not.
Because if the item features are available in the dataframe, one would need to create a new SparkDiversityEvaluation object if one wishes to compute the co-occurrence based similarities.

Reply:
Do you mean we want to only create a single SparkDiversityEvaluation object, and be able to get diversity value from two different item similarity approaches if item features are available? 
e.g.
eval = SparkDiversityEvaluation(args, ...)
diversity_item_coocurrence = eval.diversity(method="item_coocurrence")
diversity_item_similarity=eval.diversity(method="item_similarity")

Reply:
If using this approach, it doubles the code in multiple functions. It looks very confusing. Therefore I kept the previous method unchanged, i.e. only one "item_sim_measure" can be chosen for one object 
eval = SparkDiversityEvaluation(args, ...). 

"select the method for calculating similarity using an argument instead of whether item_feature_df is present or not."  -- this is implemented.

Reply:
Why would it increase the code significantly? I think you would just need to add an extra argument in some functions and just pass this argument from calling to called function. So that you could write
`als_diversity=als_eval.diversity(method="cooccurrence")`
or something similar.
And you would remove the argument from the constructor that you added in the most recent commit.

Reply:
The main thing I would like to avoid is the user creating two SparkDiversityEvaluation objects when they want to compute both types of diversity or serendipity metrics. This is more user-unfriendly than the alternative. And imagine what would happen if we added more item similarity methods: they would need to create as many objects as the methods available.

Reply:
it will looks something like 
![image](https://user-images.githubusercontent.com/10565026/132579849-ba08e4e8-63ad-4dda-aff0-d1e2839f1a41.png)

Reply:
@anargyri @gramhagen what do you think?  Should we make change to use the above method?

Reply:
Actually, you don't need to duplicate the code in this way. What I meant was to change some of the member DFs into dictionaries indexed by the similarity method. I.e. do something like this instead:
```
def user_diversity(self, item_sim_measure="item_cooccurrence_count"):
        if self.df_user_diversity[item_sim_measure] is None:
            self.df_intralist_similarity[item_sim_measure] = self._get_intralist_similarity(self.reco_df, item_sim_measure=item_sim_measure)
            self.df_user_diversity[item_sim_measure] = (
                self.df_intralist_similarity[item_sim_measure].withColumn(
                    "user_diversity", 1 - F.col("avg_il_sim")
                )
                .select(self.col_user, "user_diversity")
                .orderBy(self.col_user)
            )
        return self.df_user_diversity[item_sim_measure]
```
So the code doesn't really increase but it becomes less readable. I am fine both ways but it's a trade-off, i.e. shifting convenience from the end-user to the developer or vice versa.
@yueguoguo @miguelgfierro any opinions on this?